### ISSUE TITLE
PTY not being used in linux terminals

### ISSUE THREAD

**Issue by TotallyNotElite** [Ylianst/MeshCentral#903]
Some recent commits have stopped the usage of PTYs in meshcentral. What's the reason for this? This is a pretty big deal for me and my team.

**Comment by krayon007** [comment]
What specifically are you referring to? The terminal on linux currently uses forkpty(), so the terminal session is still running as a pseudo terminal.

 If you are referring to the usage of 'python' and 'script' from the terminal, I had to remove it, because the way it was implemented was broken. Older distributions of linux, shipped with an older version of script that did not support the same command line options, so would fail to start. (Terminal would connect, then immediately disconnect). Other versions of linux shipped with a version of script where the behavior was diferent in that it would not properly echo. I also found that the python fallback did not correctly handle process handles, resulting in hung/leaked processes, that caused the parent process to hang, because I did a waitpid to make sure the process correctly cleaned up, which caused problems for auto-update.

**Comment by TotallyNotElite** [comment]
I still can't confirm this working. Same behavior as described in my original pull request. https://github.com/Ylianst/MeshCentral/pull/677#issuecomment-557916444 This makes the terminal feature basically unusable for us, which is the reason I added that in the first place.

**Comment by krayon007** [comment]
The main problem with the python solution, is that pty.spawn() does not return until the child process exits. If the child process never exits, spawn() will never return. If the python process is killed, it still does not kill the spawned process., and on some platforms, I saw that the child process goes to 100% CPU if the python process is killed. This means if I kill the python process, and waitpid() to cleanup correctly, it never returns. 

The other issue, is that the python process seems to swallow SIGWINCH and not pass them to the child, so resize does not work on the child.  The spawn() api doesn't appear to pass me back the PID of the child, as it is a blocking call.  The proposed solution with setting a trap, does not work on many of the distributions I tested on. The only one that seemed to work, was using an IOCTL, which is what SSHd does.

If I can resolve these issues, I can put the python solution back in.

**Comment by TotallyNotElite** [comment]
I don't really care about the python solution, it was temporary anyway. There has to be __some kind__ of pty, otherwise you will get non-interactive terminals everywhere. [The solution involving `script` is also disabled](https://github.com/Ylianst/MeshCentral/blob/master/agents/meshcore.js#L1242).

> python process seems to swallow SIGWINCH

This isn't actually limited to python. It's not normal for processes to pass signals down to their children. `script` has the same problem.

>  The proposed solution with setting a trap, does not work on many of the distributions I tested on.

It also breaks whenever you switch to a new user, so not ideal anyway.

I think there's no way around properly creating a PTY using syscalls if the python and script solutions don't work.

**Comment by krayon007** [comment]
@TotallyNotElite , I almost forgot... We actually added an xoption JSON object, that the server passes to the agent. In this JSON object, we actually created a boolean for 'script' and 'python', such that if the server sets them, the agent will use script if it's detected on the agent as version 2.25 or newer, falling back to python if its not. (Which is essentially the behavior we had a few days ago). I'm not sure if @Ylianst  created a configuration to set this JSON object in the UI or not...

**Comment by TotallyNotElite** [comment]
I am aware of said variable, but the commits which added the variable doesn't seem to include a way to set said variable (if I remember correctly). In my humble opinion, script or another solution should definitely be enabled by default. In the beginning I was considering looking for another tool just because of that issue.

**Comment by Ylianst** [comment]
I will work on adding a switch to enable script/python on the web application, this said... I am consufed by this issue... We are attempting to launch the shell correctly (the same way SSH does, I don't think SSH does script/python tricks). What is not working in the current terminal that should?

**Comment by TotallyNotElite** [comment]
> the same way SSH does

No, meshcentral does weird hacks and tries to forces bash to act like its in an interactive shell using the -i parameter. The issue is that any subsequent shells that may be created are non-interactive shells, which are basically unusable.

**Comment by ryanblenis** [comment]
Not sure if this is related to this thread or needs its own issue, but just noticed that both "Terminal" and "XTerm" both yield no output (yet you can still type commands and see their responses) after a `su -s /bin/bash USERNAME` host OS Ubuntu 18.04

**Comment by TotallyNotElite** [comment]
@ryanblenis That's what this issue is about. Non interactive shells.

**Comment by krayon007** [comment]
ok, I finally root caused this, and fixed this the correct way. I found I actually did create the shell correctly, and correctly allocated a pty. The problem was, that I was redirecting stderr. 

isatty() succeeded for stdin and stdout, but not for stderr. I think bash was looking at stderr to determine if it was interactive or not. So I modified the behavior, so when you spawn a child process as interactive, I don't redirect stderr...

Everything appears to function correctly now... I can create a bash subshell without specifying -i and I can use su and switch users, and it works correctly...

This fix requires a new linux agent, so Ylian is doing regression testing, so we'll push an update soon...
